projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/llvm/arith/LLVMUAddWithOverflow.java

```diff
@@ -41,16 +41,60 @@
 
     @NodeChildren({@NodeChild(value = "left", type = LLVMExpressionNode.class), @NodeChild(value = "right", type = LLVMExpressionNode.class),
                     @NodeChild(value = "target", type = LLVMExpressionNode.class)})
+    public abstract static class LLVMUAddWithOverflowI8 extends LLVMBuiltin {
+
+        @Specialization
+        public LLVMAddress executeIntrinsic(byte left, byte right, LLVMAddress addr) {
+            final int res = (left & LLVMExpressionNode.I8_MASK) + (right & LLVMExpressionNode.I8_MASK);
+            final boolean overflow = (res & (1 << Byte.SIZE)) != 0;
+
+            LLVMMemory.putI8(addr, (byte) (res));
+            LLVMMemory.putI1(addr.getVal() + Byte.BYTES, overflow);
+            return addr;
+        }
+    }
+
+    @NodeChildren({@NodeChild(value = "left", type = LLVMExpressionNode.class), @NodeChild(value = "right", type = LLVMExpressionNode.class),
+                    @NodeChild(value = "target", type = LLVMExpressionNode.class)})
+    public abstract static class LLVMUAddWithOverflowI16 extends LLVMBuiltin {
+
+        @Specialization
+        public LLVMAddress executeIntrinsic(short left, short right, LLVMAddress addr) {
+            final int res = (left & LLVMExpressionNode.I16_MASK) + (right & LLVMExpressionNode.I16_MASK);
+            final boolean overflow = (res & (1 << Short.SIZE)) != 0;
+
+            LLVMMemory.putI16(addr, (short) (res));
+            LLVMMemory.putI1(addr.getVal() + Short.BYTES, overflow);
+            return addr;
+        }
+    }
+
+    @NodeChildren({@NodeChild(value = "left", type = LLVMExpressionNode.class), @NodeChild(value = "right", type = LLVMExpressionNode.class),
+                    @NodeChild(value = "target", type = LLVMExpressionNode.class)})
     public abstract static class LLVMUAddWithOverflowI32 extends LLVMBuiltin {
 
-        private static final int OVERFLOW_FIELD_OFFSET = 4;
+        @Specialization
+        public LLVMAddress executeIntrinsic(int left, int right, LLVMAddress addr) {
+            final int res = left + right;
+            final boolean overflow = (((~res & left) | (~res & right) | (left & right)) & (1 << (Integer.SIZE - 1))) != 0;
```

Comment:
Is there a particular reason, why you write "& (1 << (Integer.SIZE - 1))) != 0" instead of "< 0"? I used "< 0", because I thought it should work like for the signed version.